### PR TITLE
feat(protocol): send worker info (id and pid) to broker

### PR DESCRIPTION
**Pull request by bungle**
### Summary

This is a feature that sends worker info to broker. This feature is needed for future PRs to implement worker id based broker queues (to replace worker connection based queues).

### CODE REVIEW

File: lualib/resty/events/protocol.lua [hunk 1]
```diff
@@ -154,6 +167,11 @@ function _Client.connect(self, addr)
         end
     end -- subsystem == "http"
 
+    local _, err = send_frame(self, tostring(worker_id() or -1))
```
Comment (outsinre) [inline lualib/resty/events/protocol.lua:170]: In the current implementation, the broker is also the client of itself. Should we consider this situation?

For example, we can check if `self.broker_id` is equal to `worker_id()`.

**Review by outsinre — COMMENTED** [review]
We can postpone this as a standalone job.

File: lualib/resty/events/protocol.lua (hunk 1 above)
Comment (bungle) [reply lualib/resty/events/protocol.lua:170]: I don’t think that needs any special treatment.

Comment (bungle) [reply lualib/resty/events/protocol.lua:170]: And it keeps code cleaner when broker worker also connects to broker.

**Comment by bungle** [comment]
> We can postpone this as a standalone job.

What does postponing as a standalone job mean? Do you mean we can cut a release without it? I completely agree. This PR just builds basis for future improvements that I will open ”soon”.

File: lualib/resty/events/protocol.lua (hunk 1 above)
Comment (outsinre) [reply lualib/resty/events/protocol.lua:170]: Make sense.

Comment (bungle) [reply lualib/resty/events/protocol.lua]: The `or -1` is what this library is to give `privileged agent` a number that is normally a `nil`.

**Comment by chronolaw** [comment]
If it is not urgent, I intend to turn it to draft.

File: lualib/resty/events/protocol.lua [hunk 2]
```diff
@@ -83,12 +92,22 @@ function _Server.new()
...
     sock:settimeout(DEFAULT_TIMEOUT)
 
+    local data, err = _recv_frame(sock)
+    if err then
+        return nil, "failed to read worker info from worker: " .. err
+    end
+
+    local info = decode(data)
+    if not info then
+        return nil, "invalid worker info: " .. data
```
Comment (chronolaw) [inline lualib/resty/events/protocol.lua:102]: I think that `data` is unprintable.

Comment (bungle) [reply lualib/resty/events/protocol.lua]: changed this.

**Comment by chronolaw** [comment]
Question: what is the advantage of replacing connection with worker id? Could I get some explain?

**Comment by bungle** [comment]
> Question: what is the advantage of replacing connection with worker id? Could I get some explain?

Workers can crash. So this code:
```
self._clients[worker_connection]:pop()
```

if it was something like:
```
QUEUES[worker_id]:pop()
```

We could make queues survive across worker crashes (and continue writing on them). We could also simplify a lot of loops to just:
```
for i=1, WORKER_COUNT do
end
```

I think that with this library we can and probably should assume that each worker connects to a broker.

**Review by chronolaw — APPROVED** [review]

File: lualib/resty/events/protocol.lua [hunk 3]
```diff
@@ -153,6 +172,14 @@ function _Client:connect(addr)
         end
     end -- subsystem == "http"
 
+    WORKER_INFO.id = worker_id()
```
Comment (outsinre) [inline lualib/resty/events/protocol.lua:175]: LGTM. Just wonder if we can add robustness check on the return value.

**Review by outsinre — APPROVED** [review]

File: lualib/resty/events/protocol.lua [hunk 4]
```diff
@@ -83,12 +92,22 @@ function _Server.new()
 
     sock:settimeout(DEFAULT_TIMEOUT)
 
+    local data, err = _recv_frame(sock)
+    if err then
+        return nil, "failed to read worker info: " .. err
+    end
+
+    local info, err = decode(data)
```
Comment (outsinre) [inline lualib/resty/events/protocol.lua:100]: If we can add a debug message which worker is connecting?

File: lualib/resty/events/protocol.lua (hunk 3 above)
Comment (chronolaw) [reply lualib/resty/events/protocol.lua:175]: This line should be `worker_id() or -1`, for privileged worker.

Comment (bungle) [reply lualib/resty/events/protocol.lua]: Added.

File: lualib/resty/events/protocol.lua (hunk 4 above)
Comment (bungle) [reply lualib/resty/events/protocol.lua:100]: @outsinre, I think logging that belongs to `broker`, in this PR I only wanted to change `protocol.lua`.

Comment (chronolaw) [reply lualib/resty/events/protocol.lua:100]: Is this message helpful for debugging?

File: t/protocol.t [hunk 5]
```diff
@@ -33,6 +33,7 @@ __DATA__
                 end
 
                 ngx.log(ngx.DEBUG, "Upgrade: ", ngx.var.http_upgrade)
+                ngx.log(ngx.DEBUG, "Worker ID: ", conn.info.id)
```
Comment (chronolaw) [inline t/protocol.t:36]: Should we add a same test in `privileged.t` to verify if worker_id == -1 ?

File: lualib/resty/events/protocol.lua (hunk 4 above)
Comment (outsinre) [reply lualib/resty/events/protocol.lua:100]: Make sense. If we add this debug, we add to `broker`, not the `protocol`.

**Review by tzssangglass — DISMISSED** [review]
LGTM

File: t/protocol.t (hunk 5 above)
Comment (bungle) [reply t/protocol.t:36]: added

**Review by gszr — APPROVED** [review]

**Pull request merged**